### ISSUE TITLE
Error: formValues() must be used inside a React tree decorated with reduxForm()

### ISSUE THREAD

**Issue by someden** [erikras/redux-form#4299]
**Bug report**

Demo: https://stackblitz.com/edit/redux-form8?file=Form.js
But demo doesn't work because of https://github.com/erikras/redux-form/issues/4293

**Steps to reproduce**
```bash
$ npx create-react-app form
$ cd form
$ yarn add redux redux-form react-redux
# Copy code from demo index.js and Form.js
$ yarn start
```

So, about bug:
```javascript
import { reduxForm, formValues } from 'redux-form';
/* ... */
export default reduxForm({ form: 'Form' })(formValues('name')(Form));
```
After update to version 8.0.1 using `formValues` decorator results in an error:
```
Uncaught Error: formValues() must be used inside a React tree decorated with reduxForm()
```

**Comment by dagstuan** [comment]
The issue still exists in 8.0.4, working (broken) example: https://codesandbox.io/s/21m68rvozn

**Comment by erikras** [comment]
Confirmed. Will investigate tomorrow.

**Comment by tyscorp** [comment]
It looks like the consumer is being called before the provider, even though the provider is at a higher level in the tree. I have no idea how/why.

**Comment by erikras** [comment]
It's got to be something to do with using `Context.Provider` rather than just a HOC to create the context. Hmm... You'll have to put the component connected to `formValues()` deeper in the tree.

Until I can figure it out, here's a little render-prop workaround for you.

[![Edit Redux Form - Simple Example](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/8k766888mj)

**Comment by tyscorp** [comment]
Here is a version of that using recompose's [toRenderProps](https://github.com/acdlite/recompose/blob/master/docs/API.md#torenderprops) and [fromRenderProps](https://github.com/acdlite/recompose/blob/master/docs/API.md#fromrenderprops) that allows the same usage as before.

This is very strange. I wonder if it's a bug in React?

**Comment by tyscorp** [comment]
@erikras Removing the `withReduxForm` from [this line](https://github.com/erikras/redux-form/blob/755b670864c52d544d6e63c54d33844ed3772b1c/src/createReduxForm.js#L1152) fixes the issue.

**Comment by erikras** [comment]
> This is very strange. I wonder if it's a bug in React?

I assume there's no way that @gaearon can have the time to check his GH mentions, but...?

**Comment by cajoy** [comment]
@tyscorp You workaround works mostly perfectly except it lose "children" prop when you do `fromRenderProps(toRenderProps(`

**Comment by dagstuan** [comment]
Any news on this issue?

**Comment by dagstuan** [comment]
@erikras will this be fixed any time soon?

**Comment by tnrich** [comment]
Also hitting this when updating to v8. Reverting back to 7 now!

**Comment by mr-aspex** [comment]
Hi guys. 
Any updates from a redux-form team?

**Comment by jedwards1211** [comment]
It's time for an HoC-free alternative API.  HoCs need to die already.

**Comment by jedwards1211** [comment]
The following hook works inside a `reduxForm()`:

```js
import * as React from 'react'
import { useSelector } from 'react-redux'
import { ReduxFormContext } from 'redux-form/es/ReduxFormContext'
import formValueSelector from 'redux-form/es/formValueSelector'
import { mapValues } from 'lodash'

export default function useFormValues(values, deps = []) {
  const { form, sectionPrefix } = React.useContext(ReduxFormContext)
  deps = [form, sectionPrefix, ...deps]
  const mapStateToProps = React.useMemo(() => {
    const valueSelector = formValueSelector(form)
    const finalValues = sectionPrefix
      ? mapValues(values, v => `${sectionPrefix}.${v}`)
      : values
    return state => valueSelector(state, ...finalValues)
  }, deps)
  return useSelector(mapStateToProps, deps)
}
```

**Comment by jedwards1211** [comment]
If I replace `withReduxForm` with this, it works:
```js
    const ConnectedFormValues = props => (
      <FormValues {...props} _reduxForm={React.useContext(ReduxFormContext)} />
    )
    return ConnectedFormValues
```
This is why I hate HoCs and `forwardRef` (which was a bad solution to just one of HoCs' flaws).  As @tyscorp has pointed out, something within `withReduxForm` (and I'm guessing it has to do with `forwardRef`) is causing the problem.

**Comment by asazernik** [comment]
> It's time for an HoC-free alternative API. HoCs need to die already.

There's a pretty large installed base for the current API, so the bugs still need to be fixed.

**Comment by asazernik** [comment]
@tyscorp 

> I'm not sure what other effects it has, but there is a lot of console.errors when I run the test suite.

Turns out those errors hit the console in master; they are from formValues.spec.js expecting errors in certain cases, but not suppressing the log output. Not sure how to fix that, but it's out of scope. I'm testing out your suggested fix in my codebase, and will make a PR if it works.

**Comment by renatoagds** [comment]
Published in [`v8.2.5`](https://github.com/erikras/redux-form/releases/tag/v8.2.5).
